### DISCUSSION TITLE
Visualizing 2D and 3D vector and tensor data from firedrake in python

### DISCUSSION THREAD

**Discussion by Arijit-Majumdar** [firedrakeproject/firedrake#3443] [Firedrake support]
Hi, I am trying to solve linear thermoelasticity problem in firedrake for both 2D and 3D geometries. My function space for displacement and solution function is as follows
`V = VectorFunctionSpace(mesh, "CG", 1)`
`u_soln = Function(V)`

Afterwards I am calculating the stress and the strain rate tensor using the following functions
`def epsilon(u, Temp):`
 ` return 0.5*(grad(u) + grad(u).T) - alpha*Temp*Id`

`def sigma(u, Temp):`
`  return lambda_*div(u)*Id + 2*mu*epsilon(u, Temp)`

I wanted to know what is the syntax for plotting contours of different components of the displacement (u_soln) and stress/strain tensor in python? I tried the following lines for the vertical displacement but it gave me an error

`tricontourf(u_soln[1], axes=axes)`

Also, for a 3D domain, how do I get the displacement or stress/strain at a given plane (e.g. z = 0.5) and make a contour plot of specific components?

Thanks

**Comment by connorjward** [comment]
I don't know how to visualise a plane, but to visualise a component of a vector `Function` you want to use `tricontourf(u_soln.sub(1), ...)`.

For the plane I guess you could build the array you wanted using [point evaluation](https://www.firedrakeproject.org/point-evaluation.html#point-evaluation) and then plot it with ordinary matplotlib routines.

**Reply by Arijit-Majumdar** [reply]
Thanks! Does the sub work with tensor as well. For example I am evaluating the strain rate as 
`epsilon = 0.5*(grad(u) + grad(u).T)`
So for 2D problem `epsilon` will be 2 x 2 tensor at each mesh point. So if I want to plot `epsilon[1][1]` how should I use `sub`? The documentation says that `sub` takes only 1 argument.

**Reply by connorjward** [reply]
From reading the code it seems like you want to use `epsilon.sub(3)`. I.e. flatten the multiindex to `epsilon.sub(1*2 + 1)`. I agree this isn't at all clear.

**Comment by dham** [comment]
For serious plotting you should output to VTU and use paraview. This will give you a lot of plotting options beyond those which are in the basic built-in plotting functionality. It will also be vastly faster (especially in 3D).

**Reply by Arijit-Majumdar** [reply]
Thanks for the suggestion. Actually we are using Firedrake for teaching students in a Finite element course and wanted to keep all the coding and analysis in python for the ease of the students.
